libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -47,10 +48,10 @@ type
     handler*: seq[TopicHandler]
 
   PubSub* = ref object of LPProtocol
+    switch*: Switch                                 # the switch used to dial/connect to peers
```

Comment:
instead of the switch here, how about just taking a `DialCallback` closure? that'll disconnect pubsub from the switch implementation, allowing it to be more "standalone" which can help testing and break cyclic dependencies - in particular, it would enable the simple creation of tests that work the failure path

Reply:
not a bad idea 👍